### PR TITLE
New optimization functions 

### PR DESCRIPTION
**Pull request by antoniomezzacapo**
Added 3 new functions and styled comments. 

## Description
- group_paulis groups a Pauli Hamiltonian into TPB sets. 
- print_pauli_list_grouped displays the grouped Hamiltonian
- eval_hamiltonian is a single function that handles multiple cases for observable evaluation:
   - 1 shot (perfect evaluation) : the Hamiltonian (or observable) can be provided as a 1d array (classical case, diagonal observable), a matrix, or a pauli_list
    - N!=1 shots : the Hamiltonian has to be provided as a pauli list grouped into TPB sets via group_paulis

## Motivation and Context
Efficient observable evaluation
Emulation of experimental sampling

### CODE REVIEW

File: qiskit/tools/apps/optimization.py [hunk 1]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+    if shots == 1:
+
+        if type(hamiltonian) is not list: # Hamiltonian is not a pauli_list grouped into tpb sets
+
+            circuit=['c']
+            Q_program.add_circuit(circuit[0],input_circuit)
+            result = Q_program.execute(circuit,device,shots=shots,silent=True)
+            quantum_state = result.get_data(circuit[0])['quantum_state']
+            
+            if hamiltonian.shape[0] == 1 or np.shape(np.shape(np.array(hamiltonian))) == (1,): # diagonal Hamiltonian represented by 1-d array
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: For long comments, you can put them above the commented code.

File: qiskit/tools/apps/optimization.py [hunk 2]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 3]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            circuit=['c']
+            Q_program.add_circuit(circuit[0],input_circuit)
+            result = Q_program.execute(circuit,device,shots=shots,silent=True)
+            quantum_state = result.get_data(circuit[0])['quantum_state']
+            
+            if hamiltonian.shape[0] == 1 or np.shape(np.shape(np.array(hamiltonian))) == (1,): # diagonal Hamiltonian represented by 1-d array
+                
+                energy = np.sum(hamiltonian*np.absolute(quantum_state) ** 2)
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 4]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            result = Q_program.execute(circuit,device,shots=shots,silent=True)
+            quantum_state = result.get_data(circuit[0])['quantum_state']
+            
+            if hamiltonian.shape[0] == 1 or np.shape(np.shape(np.array(hamiltonian))) == (1,): # diagonal Hamiltonian represented by 1-d array
+                
+                energy = np.sum(hamiltonian*np.absolute(quantum_state) ** 2)
+        
+            elif (hamiltonian.shape[0] == hamiltonian.shape[1]): # Hamiltonian represented by square matrix 
+                energy = np.inner(np.conjugate(quantum_state),np.dot(hamiltonian,quantum_state))
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 5]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+        else: # Hamiltonian represented by a Pauli list
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 6]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            # Trial circuit w/o the final rotations 
+            circuits_labels.append('circuit_label0')
+            Q_program.add_circuit(circuits_labels[0],circuits[0])
+    
+            # Execute trial circuit with final rotations for each Pauli in hamiltonian and store from circuits[1] on
             
+            q = QuantumRegister("q", int(np.log2(len(hamiltonian))))
+            i=1
+            for p in hamiltonian:
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 7]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+                        circuits[i].z(q[j])
+                    elif p[1].v[j] == 1 and p[1].w[j] == 1:
+                        circuits[i].y(q[j])
+
+                circuits_labels.append('circuit_label'+str(i))
+                Q_program.add_circuit(circuits_labels[i],circuits[i])
+                i+=1
+            
+            result=Q_program.execute(circuits_labels, device, shots=shots, silent=True)
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 8]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            quantum_state_0=result.get_data(circuits_labels[0])['quantum_state'] # no Pauli final rotations
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: Put the comment in the line above

File: qiskit/tools/apps/optimization.py [hunk 9]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            i=1
+            for p in hamiltonian:
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 10]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            result=Q_program.execute(circuits_labels, device, shots=shots, silent=True)
+
+            quantum_state_0=result.get_data(circuits_labels[0])['quantum_state'] # no Pauli final rotations
+            i=1
+            for p in hamiltonian:
+                
+                quantum_state_i=result.get_data(circuits_labels[i])['quantum_state'] # final rotations of (i-1)-th Pauli 
+                energy += p[0]*np.inner(np.conjugate(quantum_state_0),quantum_state_i)
+                i+=1
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 11]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            quantum_state_0=result.get_data(circuits_labels[0])['quantum_state'] # no Pauli final rotations
+            i=1
+            for p in hamiltonian:
+                
+                quantum_state_i=result.get_data(circuits_labels[i])['quantum_state'] # final rotations of (i-1)-th Pauli 
+                energy += p[0]*np.inner(np.conjugate(quantum_state_0),quantum_state_i)
+                i+=1
+    
+    else: # finite number of shots and hamiltonian grouped in tpb sets 
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 12]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+        circuits=[]
+        circuits_labels=[]
+        n = int(len(hamiltonian[0][0][1].v))
+        q = QuantumRegister("q", n)
+        c = ClassicalRegister("c", n)
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 13]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+        i=0
+        for tpb_set in hamiltonian:
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 14]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+        c = ClassicalRegister("c", n)
+
+        i=0
+        for tpb_set in hamiltonian:
+  
+            circuits.append(copy.deepcopy(input_circuit))
+            circuits_labels.append('tpb_circuit_'+str(i))
+
+            for j in range(n):
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 15]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+                if tpb_set[0][1].v[j] == 0 and tpb_set[0][1].w[j] == 1: # Measure X 
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 16]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            circuits_labels.append('tpb_circuit_'+str(i))
+
+            for j in range(n):
+
+                if tpb_set[0][1].v[j] == 0 and tpb_set[0][1].w[j] == 1: # Measure X 
+                
+                    circuits[i].h(q[j])
+                
+                elif tpb_set[0][1].v[j] == 1 and tpb_set[0][1].w[j] == 1: # Measure Y
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 17]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+                
+                elif tpb_set[0][1].v[j] == 1 and tpb_set[0][1].w[j] == 1: # Measure Y
+                
+                    circuits[i].s(q[j]).inverse()
+                    circuits[i].h(q[j])
+
+                circuits[i].measure(q[j], c[j])
+
+            Q_program.add_circuit(circuits_labels[i],circuits[i])
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 18]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            i+=1
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 19]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+
+        result=Q_program.execute(circuits_labels, device, shots=shots, silent=True)
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 20]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+        for j in range(len(hamiltonian)):
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 21]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+            for k in range(len(hamiltonian[j])):
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 22]
```diff
@@ -136,61 +153,233 @@ def Measure_pauli_z(data, pauli):
...
+                energy += hamiltonian[j][k][0] * Measure_pauli_z(result.get_counts(circuits_labels[j]),hamiltonian[j][k][1])
+
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: nit: remove extra line

File: qiskit/tools/apps/optimization.py [hunk 23]
```diff
@@ -17,78 +17,94 @@
...
-from qiskit.tools.qi.pauli import Pauli, label_to_pauli
-
-
-def SPSA_optimization(obj_fun, initial_theta, SPSA_parameters, max_trials, save_steps = 1,last_avg=1):
-    """Minimize the obj_fun(controls).
-
-    initial_theta = the intial controls
-    SPSA_parameters = the numerical parameters
-    max_trials = the maximum number of trials
+import copy
```
Comment (atilag) [inline qiskit/tools/apps/optimization.py]: Can you move this import behind: import os ?  The recommendation is to group system imports together.

**Comment by dtmcclure** [comment]
Looks like this update has not made it into the version of qiskit available on pip. How often does that version get synced with the latest version from GitHub?

**Comment by atilag** [comment]
Oh, It's instantaneous! :) I just need to take these changes and merge them into our release version. I just didn't know we wanted to create a new release version with these changes. I have to port them to our release branch before creating a new PIP package, so it won't be that fast this time.

**Comment by dtmcclure** [comment]
Oh I see, so PIP syncs to the latest release branch on GitHub, which is different from master. Whether/when to create a new release version with these changes is for someone other than me to decide though :-). Now at least I understand how it works!

**Pull request merged**